src/charm.py

```diff
@@ -1266,21 +1413,27 @@ def _initialise_replica_set(self, event: StartEvent) -> None:
                 )
                 event.defer()
                 self.status.set_and_share_status(
-                    WaitingStatus("waiting to initialise replica set")
+                    WaitingStatus("[ERR1] waiting to initialise replica set")
                 )
                 return
             except PyMongoError as e:
                 logger.error("Deferring on_start since: error=%r", e)
                 event.defer()
                 self.status.set_and_share_status(
-                    WaitingStatus("waiting to initialise replica set")
+                    WaitingStatus("[ERR2] waiting to initialise replica set")
                 )
                 return
 
             # replica set initialised properly and ready to go
             self.db_initialised = True
+            self.lock_hash = generate_lock_hash()
```

Comment:
I haven't read this yet, but I assume it sets the lock hash across all units a a file?

Reply:
The lock_hash attribute is a descriptor.
Writing to `self.lock_hash` will execute the `__set__` function of the `LockHashHandler` class which does two things:
 * Write in the file of the unit
 * Write the lock hash in the secret if we're the leader.

So what happens here is that the leader will generate the lock hash once and write it to it's fs and in the secret.
And then the other nodes will fetch it from the secret and write it on their fs during the `RelationJoined` event.